### ISSUE TITLE
BED file is empty after running wig2bed on WigFix overnight, had to terminate

### ISSUE THREAD

**Issue by RohanParikh00** [bedops/bedops#255]
I am trying to convert a wigFix file to a bed file using wig2bed; I have done this before using the UCSC wig files that are per-chromosome, but never when the wig files are not broken up by chromosome (e.g. the wigFix file from https://hgdownload.soe.ucsc.edu/goldenPath/xenTro9/phyloP11way/). This is what the beginning of the wigFix file looks like (it is a multi-section fixedStep file):
<img width="412" alt="wigfixfile" src="https://user-images.githubusercontent.com/20759333/118011038-9333c880-b31d-11eb-9c09-9adae5670a5c.png">

I attempted to convert it to a bed file using `wig2bed --multisplit < xenTro9.phyloP11way.wigFix > xenTro9.bed` but the code is running for hours on a cluster and not writing to the bed file, so I had to terminate. Any help would be appreciated!

**Comment by alexpreynolds** [comment]
If you are using `--multisplit=<val>`, you need to specify a value `<val>` or that option will not work and `wig2bed` will hang.

I did a quick test, and I'm not noticing any conversion problems. 

But after trying to convert this file, I did notice that it is very large and some other options are likely necessary.

Here's how I retrieved the file:

```
$ wget -qO- https://hgdownload.soe.ucsc.edu/goldenPath/xenTro9/phyloP11way/xenTro9.phyloP11way.wigFix.gz > xenTro9.phyloP11way.wigFix.gz
```

This the version of `convert2bed` (`wig2bed`) that I am running:

```
$ convert2bed --version
convert2bed
  version:  2.4.39
  author:   Alex Reynolds
```

Here is the conversion command I am using to generate _unsorted_ output:

```
$ wig2bed --multisplit=foo --do-not-sort < <(gunzip -c xenTro9.phyloP11way.wigFix.gz) >  xenTro9.phyloP11way.unsorted.bed
```

Here is what output looks like:

```
$ head xenTro9.phyloP11way.unsorted.bed
chrUn_NW_016683437v1	32	33	foo.1-1	0.301000
chrUn_NW_016683437v1	33	34	foo.1-2	0.297000
chrUn_NW_016683437v1	34	35	foo.1-3	-0.809000
chrUn_NW_016683437v1	35	36	foo.1-4	0.297000
chrUn_NW_016683437v1	36	37	foo.1-5	0.301000
chrUn_NW_016683437v1	37	38	foo.1-6	0.301000
chrUn_NW_016683437v1	38	39	foo.1-7	0.301000
chrUn_NW_016683437v1	39	40	foo.1-8	0.301000
chrUn_NW_016683437v1	40	41	foo.1-9	-0.823000
chrUn_NW_016683437v1	41	42	foo.1-10	0.301000
```

I would ask that you test this same command on your end: i.e., 1) add a string to `--multisplit`; 2) add `--do-not-sort`; 3) pipe to `head` to see what the output looks like.

If this works, then a problem is that your system is potentially running out of memory.

The original compressed wig file you are using will blow up a lot on expansion and conversion.

After conversion, there is a sorting step run internally. This sorting step requires system memory, unless options are specified to use a disk-based sort. If your host runs out of memory with the default sorting step, the process and your host will likely hang.

If that `--do-not-sort` test works, then the next thing I would suggest that you do is run `free -m` or similar (`vm_stat` on Mac OS X) to see how much available system memory you have, when you are converting the file.

Then remove `--do-not-sort` and add, for example: `--max-mem=1G --sort-tmpdir=${PWD}`. 

If running `free -m` or similar shows that you have more than 1Gb of free system memory, then you can replace `1G` with a larger value, which will make sorting go faster, e.g. `2G`, `4G`, etc. This setting depends on your host.

The `--sort-tmpdir` variable sets where intermediate sort files are placed. This needs to be specified if you are using `--max-mem`. If you're running out of system memory, then you want to change this to whatever non-`/tmp` directory you may be working in, which has sufficient disk space to transiently store intermediate sort data.

The following commands are functionally identical:

```
$ wig2bed  --max-mem=1G --sort-tmpdir=${PWD} < <(gunzip -c in.wig.gz) > out.bed
```

And:

```
$ wig2bed  --do-not-sort < <(gunzip -c in.wig.gz) | sort-bed --max-mem 1G --tmpdir ${PWD} - > out.bed
```

Which you use is your preference.

Converting this file will take a while using disk-based sort, maybe 20-40 minutes or so. You'll probably need about 30-40 x 500Mb of disk space for intermediate sort data, and the same for storing the final output BED file. 

If you're doing set operations, you might look at compression to Starch, which will take longer to make but also be a much smaller file than the BED file — closer to the original gzip file from UCSC than the whopper that the uncompressed BED will be.

**Comment by RohanParikh00** [comment]
Thank you for your detailed response. This worked perfectly, and I was able to generate the sorted BED file!

**Comment by rsharris** [comment]
@alexpreynolds I experienced similar behavior, resolved by adding --do-not-sort. Except in my case wig2bed finished, reported no errors, and produced no output.

Specifically, I had an ≈5.3Gbyte wig file that, AFAIK, was all variableSteps.  I used
```
wig2bed --multisplit=section \
  < appaloosa.wigout.wig \
  > appaloosa.mappability.bed
```
wig2bed ran for about 10 minutes but the output was empty. No error was reported. The same command had worked OK on a much smaller input.

After reading RohanParikh00's issue I added `--do-not-sort`. This run took about the same duration and produced a 6.8GByte output.  I didn't try any of the other suggestions because for my downstream use doesn't care that the .bed is unordered.

I'm quite surprised by this behavior. I was running on a slurm node that should have had 16G. I'd expect if something ran out of memory it would report an error (and fail gracefully).

**Comment by alexpreynolds** [comment]
@rsharris Thanks for the report. 

I suspect the problem might lie elsewhere; if a Slurm job attempts to use more memory than it is allocated, the job scheduler should cancel that job regardless of what `convert2bed` or `sort-bed` or other processes do, and you should typically get an email from the scheduler (or some other notification) noting that this happened due to lack of available memory.

If it is possible to share your Wiggle output somewhere that I can download locally, I'd be interested to investigate. My guess is there is something unusual or edge-case about your input, and the `convert2bed` Wiggle parser gets stuck in a loop in trying to translate it. It might be valid input and something I need to patch, or it could be something else. Please let me know if I can help further.

**Comment by rsharris** [comment]
Thanks @alexpreynolds . I'll look for a place to make the input available later today.

I do have my slurm jobs set to email me if there's a failure. So yeah, having gotten no such email for this job would suggest memory wasn't the issue. But the job finished, which would seem to preclude the stuck in a loop theory. (My slurm jobs have an `echo "job has finished"` as its final line, to make it simple to conclude whether everything else in a job has run to completion).

**Comment by rsharris** [comment]
You should be able to access the wig file via this link:
https://pennstateoffice365-my.sharepoint.com/:u:/g/personal/rsh915_psu_edu/EVHAwrB3T-9JsyGIftZynbwBs2imSOIYusBtJNpV7CVmlQ?e=tCk9g9
File size is 5,654,710,690 bytes. sha256 should be 98abc2a714c6a9f1e900c8e909d489885c92d8f02c7e1c0769396a429ec63507 .

(I'm about to post the same file in issue 239)

The command that failed was `time wig2bed  --multisplit=section < example.wig > example.bed`. Submitted as a slurm job on my machine, it took about 10 minutes, finished without reporting any error, and example.bed was empty.

The input was created by the program gem-mappability, on a diploid 6Gbp genome assembly. gem-mappability is not a supported program, but it is prescribed by this tool that I am trying to run: https://github.com/makovalab-psu/AmpliCoNE-tool . Specifically, I'm trying to accomplish the use described under "usage with other reference genomes" at the readme for that tool.

**Comment by alexpreynolds** [comment]
Thanks for sharing this file. I was able to convert this file locally, without any special options.

If `sort-bed` exits early with a non-zero exit code (i.e., sorting failed), then the following code is triggered that writes an error message to standard error:

https://github.com/bedops/bedops/blob/master/applications/bed/conversion/src/convert2bed.c#L6765-L6774

It might be worthwhile to look at the system error log output from Slurm for this job, or related jobs.

In the meantime, I added a warning message to use `--max-mem` and `--sort-tmpdir` if post-conversion sorting is needed and the input Wiggle data are too large to fit in available system memory (commit 6b5817a6). If sorting is not needed, then you can instead use `--do-not-sort`, as you have already done.

**Comment by rsharris** [comment]
Interesting. I've checked the stderr log that slurm collected and there's no sign on an error. Nor of the error message that would be produced by the code you linked. Strange.

My slurm script has `set -e` before the wig2bed command, so that any program returning an error should halt the script. It also has `set -x` so that commands are echoed to stderr. And after wig2bed completes, the script takes the output from wig2bed and passes it through a short awk script. And `wig2bed` was run as `time wig2bed`. So what I see in the log file is that wig2bed was run and took about 10 minutes. After which the awk command was run.

After the job finished, I saw that the output from wig2bed was empty.

There's no "Error: Stage [...] failed .." in the log. And the fact that the awk command was executed (with `set -e` active) says 
wig2bed didn't return an error.

The only other possibility I can think of is that I was running version 2.4.39, installed via conda. Perhaps that version behaves differently? And I don't recall testing whether `set -e` works the way it is supposed to, in the presence of conda. Seems like grasping at straws though.

From my perspective, I eventually got the result I wanted using the doughnutsort option ;). So this isn't holding me up.

Thanks.
